### PR TITLE
SITL:  The range finder and new mode in SITL

### PR DESCRIPTION
**Pull request by alexbuyval**
Hi, this PR:

* makes it possible to use the range finder in SITL;
* contains the new SITL mode for ArduCopter simulation in ROS/Gazebo.

For details see [discussion](https://groups.google.com/forum/#!topic/drones-discuss/sgGNg7IpqN8)

### CODE REVIEW

**Comment by rmackay9** [comment]
Tridge and I are trying to merge in PRs more quickly than we have in the past so hopefully we will get to this soon.
The changes generally look good to me so far.

File: libraries/AP_HAL_SITL/sitl_rangefinder.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    state.instance = 0;
+    state.distance_cm = altitude*100;
+    state.status = RangeFinder::RangeFinder_Good;
+
+    //fprintf(stdout, "2. altitude = %f \n", altitude);
+
+    _range_finder->setHIL(0, state);
+
+//     uint32_t now = hal.scheduler->millis();
+//     // add delay
```
Comment (magicrub) [inline libraries/AP_HAL_SITL/sitl_rangefinder.cpp]: Why so much commented code here?

Comment (alexbuyval) [inline libraries/AP_HAL_SITL/sitl_rangefinder.cpp]: I was going to add a delay for range finder model. But I have not decided yet which place is the best for it (Gazebo pligin or ardupilot). I can remove commented code for now and add the delay later.

Comment (lucasdemarchi) [inline libraries/AP_HAL_SITL/sitl_rangefinder.cpp]: yeah... don't add commented out code like this.

**Comment by alexbuyval** [comment]
I have removed the commented code.

**Comment by magicrub** [comment]
looks like it needs to be rebased onto master before a merge is possible.

**Comment by alexbuyval** [comment]
I have rebased my branch.

**Comment by squilter** [comment]
I think Gazebo might be the future of SITL, and this progress looks great!

@alexbuyval I see that a modified version of mavros is required to get this to work.  What if I just want to connect using a normal gcs.  How does that work?

**Comment by alexbuyval** [comment]
@squilter My modified version of mavros contains only two minor changes:
The first is a settings for Saitek ST290Pro joystick. The second is a my versions publishes control messages to 'mavros/rc/override_joy' topic instead of 'mavros/rc/override'. It is necessary to combine the control from joystick and automatic control which I used in my example. So if you use own control node (or only joystick control) you can use the normal mavros.

**Comment by squilter** [comment]
Has anybody taken another look at this PR yet?  @tridge @rmackay9 Is this PR maybe something to discuss on the dev call?

**Comment by squilter** [comment]
@alexbuyval I am getting a compile error.

```
In file included from /home/squilter/projects/tmp/ardupilot/libraries/AP_HAL_SITL/SITL_cmdline.cpp:25:0:
/home/squilter/projects/tmp/ardupilot/libraries/SITL/SIM_arducopter_sitl_ros.h:31:1: error: expected class-name before '{' token
 {
 ^
compilation terminated due to -Wfatal-errors.
../mk/build_rules.mk:27: recipe for target '/tmp/ArduCopter.build/libraries/AP_HAL_SITL/SITL_cmdline.o' failed
make: *** [/tmp/ArduCopter.build/libraries/AP_HAL_SITL/SITL_cmdline.o] Error 1
/home/squilter/projects/ardupilot/Tools/autotest/sim_vehicle.sh: Build failed
```

**Comment by alexbuyval** [comment]
@squilter Have you used the last commit? I have rebased this branch yesterday and I have not got such error.

**Comment by vooon** [comment]
@alexbuyval perhaps you may use MANUAL_CONTROL for joystick. But seems that only antenna tracker handle it...

**Comment by alexbuyval** [comment]
@vooon The control by joystick is not aim of this SITL mode. It is just example.  For instance, now I use 'GUIDED' mode for automatic control without joystick.

**Comment by squilter** [comment]
Ah, got it.  Looks like I pulled right before you added that last commit

For controlling this in auto mode without joystick, can I use this without any of your modified forks?  I think gazebo sitl would be much more popular if the needed files are all in master branch of whatever projects are used.

**Comment by alexbuyval** [comment]
@squilter In fact, there are only one required my repository. It is ['arducopter_sitl_ros' repository](https://bitbucket.org/alexbuyval/arducopter_sitl_ros) which connects 'ArduCopter' with 'ROS/Gazebo'. Other my forks are used for example of control. For instance, my fork of 'rotors_simulator' contains model of ultrasonic sensor and model of quadrotor with 450 mm frame. So if you are not going to use it you can use original repository. The same situation is with 'mavros' repository. As I wrote early, my fork of mavros just publishes control messages in other topic. 
However, I think that a user should be familiar with ROS to use this SITL mode.

**Comment by squilter** [comment]
Do you think it would be a good idea to get the code from arducopter_sitl_ros into this repo?

**Comment by alexbuyval** [comment]
@squilter The 'arducopter_sitl_ros' is ROS package. Typically, the ROS user put all packages into ROS workspace, so if we will have this repository into 'ardupilot' folder, the ROS will be not able to find it. But maybe there are approaches to solve this issue.

**Comment by squilter** [comment]
I got it working now, so I can test.  Running `sim_vehicle.sh -f arducopter_sitl_ros --console` opens gazebo with a quadcopter on the ground.  But this command does not give me mavproxy or a console.  How can I open mavproxy from here?

**Comment by squilter** [comment]
I figured out why.  `fdmUDP` is crashing because it can't find the mavros.setpoint.get_pub_attitude_throttle.  Sure enough, this function exists in mavros-master, but it does not exist in the fork that your directions told me to use.

**Comment by squilter** [comment]
After running `sim_vehicle.sh -f arducopter_sitl_ros --console`, it starts up fine.  Gazebo starts, and the second terminal starts.  But the --console doesn't appear anywhere, and the first shell never turns into mavproxy.

**Comment by alexbuyval** [comment]
@squilter The mavproxy console does not run in this SITL mode. You should use the 'mavros' package and ROS topics to control.

**Comment by squilter** [comment]
I understand that the proper way is to use mavros, but I was hoping that I could also use mavproxy or something else.  The wiki page says to run `sim_vehicle.sh -f arducopter_sitl_ros --console`, so I thought a console would also appear.

**Comment by tridge** [comment]
sorry for not doing a proper review of this PR yet, I'll get to it soon!

**Comment by nicolaerosia** [comment]
ping

**Comment by tridge** [comment]
@alexbuyval,
After an awful delay I finally had a look at this PR. My apologies it took so long.
The main thing I think you may not know is that SITL already supports a rangefinder and has done so for a long time. See the docs here:
http://ardupilot.org/dev/docs/using-sitl-for-ardupilot-testing.html#adding-a-virtual-rangefinder
The existing support just sets up an analog pin for a simulated analog rangefinder. That means no special code is needed in AP_RangeFinder.
It also has the advantage that it can use terrain data and it corrects for attitude.
The other part of this PR is adding a new ROS SITL backend. Is that still needed or is the existing SIM_Gazebo the same thing?

**Comment by khancyr** [comment]
It is not the same that the current SIM_Gazebo : it use ROS to connect to gazebo.
But it will be overtake by https://github.com/ArduPilot/ardupilot_wiki/pull/480

**Comment by tridge** [comment]
@khancyr so should I close this PR? I'm perfectly happy to have two different gazebo SITL backends if they serve a different use case.

**Comment by khancyr** [comment]
@tridge for me it could be closed . It need a rework to work with lastest Rotor simulator.

**Comment by magicrub** [comment]
Closing. this seems stale and, by the looks of the comments, needs work before it can be merged. If you still want this go ahead and and open it back up or open another pull request rebased on master,

**Pull request closed without merging**